modules/twinkleblock.js

```diff
@@ -0,0 +1,870 @@
+      { label: 'Spam/advertising-only account', value: 'uw-soablock' },
+      { label: 'Unapproved bot', value: 'uw-botblock' },
+      { label: 'Violating the three-revert rule', value: 'uw-3block' }
+    ]
+  },
+  {
+    label: 'Username violations',
+    list: [
+      { label: 'Bot username', value: 'uw-botublock' },
+      { label: 'Memorial username soft block', value: 'uw-memorialblock' },
+      { label: 'Promotional username, hard block', value: 'uw-spamublock' },
+      { label: 'Promotional username, soft block', value: 'uw-softerblock' },
+      { label: 'Similar username soft block', value: 'uw-ublock-double' },
+      { label: 'Username violation, soft block', value: 'uw-ublock' },
+      { label: 'Username violation, hard block', value: 'uw-uhblock' },
+      { label: 'Username impersonation hard block', value: 'uw-uhblock-double' },
+      { label: 'Username represents a famous person, soft block', value: 'uw-ublock-famous' },
+      { label: 'Username represents a non-profit, soft block', value: 'uw-causeblock' },
+      { label: 'Username violation, vandalism-only account', value: 'uw-vaublock' }
+    ]
+  },
+  {
+    label: 'Templated reasons',
+    list: [
+      { label: 'anonblock', value: 'anonblock' },
+      { label: 'anonblock - likely a school', value: 'anonblock - school' },
+      { label: 'blocked proxy', value: 'blocked proxy' },
+      { label: 'CheckUser block', value: 'CheckUser block' },
+      { label: 'checkuserblock-account', value: 'checkuserblock-account' },
+      { label: 'checkuserblock-wide', value: 'checkuserblock-wide' },
+      { label: 'colocationwebhost', value: 'colocationwebhost' },
+      { label: 'oversightblock', value: 'oversightblock' },
+      { label: 'school block', value: 'school block' },
+      // { label: 'rangeblock', value: 'rangeblock' },
+      { label: 'tor', value: 'tor' },
+      { label: 'webhostblock', value: 'webhostblock' }
+    ]
+  }
+];
+
+Twinkle.block.callback.filtered_block_groups = function twinkleblockCallbackFilteredBlockGroups(show_template) {
+  return $.map(Twinkle.block.blockGroups, function(value) {
+    var list = $.map(value.list, function(v) {
+        var blockSettings = Twinkle.block.blockPresetsInfo[v.value];
+        var registrationRestrict = blockSettings.forRegisteredOnly ? Twinkle.block.isRegistered : (blockSettings.forAnonOnly ? !Twinkle.block.isRegistered : true);
+        if (!blockSettings.excludeFromTemplates && registrationRestrict) {
+          return {
+            label: (show_template ? '{{' + v.value + '}}: ' : '') + v.label,
+            value: v.value
+          };
+        }
+      });
+    if (list.length) return {
+        label: value.label,
+        list: list
+      };
+  });
+};
+
+Twinkle.block.callback.change_preset = function twinkleblockCallbackChangePreset(e) {
+  var $form = $(e.target.form), key = $form.find('[name=category]').val();
+  if (!key) return;
+
+  $form.find('[name=sub_group]').val(key);
```

Comment:
Well, not quite. With Block we have separate fieldsets for the warn and block options, so we keep the same field names in the warn fieldset so we can just pass it to warn to do the actual warning. All that's happening on this line is if a preset is chosen it will choose the appropriate template for you along with the associated block options. I'm not sure if that clears it up or if your concern goes beyond that. The "nasty transform" of the `blockGroups` in Warn (line 1224) is because unfortunately Warn does not utilize the `.append({ type: 'select', list: [...array of option objects...] })` functionality Morebits provides, and instead builds the `<optgroup>`s and `<option>`s from it's own data structure format, one that's not easy to manage given how many properties we have on the block templates. Hopefully that all makes sense. Essentially my foremost concern with implementation was to avoid repetition and share functionality where possible.